### PR TITLE
[faster-rcnn] add softmaxwithcriteria

### PR DESCRIPTION
**Pull request by SeaOfOcean**
## What changes were proposed in this pull request?

softmaxwithcriterion

## How was this patch tested?

unit test

### CODE REVIEW

**Comment by SeaOfOcean** [comment]
pr validate pass

**Comment by JerryYanWan** [comment]
Can fast rcnn reuse the CrossEntropyCriterion by transposing the input and target Tensor ?

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SoftmaxWithCriterion.scala [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    var count = 0
+    i = 0
+    while (i < outerNum) {
+      var j = 0
+      while (j < innerNum) {
+        val curTarget = ev.toType[Int](labelData(i * innerNum + j))
+        if (ignoreLabel.isEmpty || ignoreLabel.get != curTarget) {
+          assert(curTarget >= 0 && curTarget < nClasses,
+            s"curTarget $curTarget is out of range 0 to ${ nClasses - 1 } ")
+          loss = ev.minus(loss,
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SoftmaxWithCriterion.scala]: please try to wrap this piece of code into a method and don't use generic inside.

File: dl/src/main/scala/com/intel/analytics/bigdl/nn/SoftmaxWithCriterion.scala [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+  @transient var outerNum = 0 // batchsize
+
+  @transient var innerNum = 1
+
+  @transient var nClasses = 2
+
+  override def updateOutput(input: Tensor[T], target: Tensor[T]): T = {
+    // input: batchsize * softmaxAxis * height * width
+    // for this example, (1, 2, 90, 4) if we have 1 image batch, 2 classes, 9*10*4 anchors
```
Comment (yiheng) [inline dl/src/main/scala/com/intel/analytics/bigdl/nn/SoftmaxWithCriterion.scala]: is this a special loss function for FastRCNN?

**Comment by SeaOfOcean** [comment]
this criterion corresponds to caffe SoftmaxWithLossLayer http://caffe.berkeleyvision.org/doxygen/classcaffe_1_1SoftmaxWithLossLayer.html

**Comment by yiheng** [comment]
LGTM
